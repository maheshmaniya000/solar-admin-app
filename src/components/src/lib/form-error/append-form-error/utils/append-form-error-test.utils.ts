import { ElementRef, Type } from '@angular/core';
import {
  AbstractControl,
  AsyncValidatorFn,
  ValidationErrors
} from '@angular/forms';
import { SpectatorDirective } from '@ngneat/spectator/jest';
import { isEmpty } from 'lodash-es';
import { Observable, of } from 'rxjs';
import { delay } from 'rxjs/operators';

import { AppendFormErrorDirective } from '../append-form-error.directive';

export class AppendFormErrorTestUtils {
  static readonly asyncRequired: (
    delayInMilliseconds: number
  ) => AsyncValidatorFn =
    (delayInMilliseconds: number) =>
    (control: AbstractControl): Observable<ValidationErrors> =>
      isEmpty(control.value)
        ? of({ required: true }).pipe(delay(delayInMilliseconds))
        : of(null);

  static expectDirectiveToBePresentOn<D extends AppendFormErrorDirective>(
    element: Element,
    type: Type<D>,
    spectator: SpectatorDirective<AppendFormErrorDirective>
  ): void {
    const appendFormErrorDirectiveElement: Element = spectator.query(type, {
      read: ElementRef
    }).nativeElement;
    expect(appendFormErrorDirectiveElement).toBe(element);
  }
}
