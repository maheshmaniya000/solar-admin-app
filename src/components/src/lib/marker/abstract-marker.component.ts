import {
  Component,
  ElementRef,
  EventEmitter,
  HostBinding,
  HostListener,
  Input,
  Output
} from '@angular/core';

import { Vector2 } from '@solargis/types/pvlib-api';

@Component({ template: '' })
export abstract class AbstractMarkerComponent {
  @HostBinding('class.selected') @Input() selected = false;
  @Output() selectedChange = new EventEmitter<void>();

  constructor(readonly elementRef: ElementRef<HTMLElement>) {}

  @HostListener('click') onClick(): void {
    this.selectedChange.emit();
  }

  getPivotPoint(): Vector2 {
    const element = this.elementRef.nativeElement;

    return {
      u: element.offsetWidth / 2,
      v: element.offsetHeight
    };
  }
}
