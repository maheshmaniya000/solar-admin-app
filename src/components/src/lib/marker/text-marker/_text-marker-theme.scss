@use '~@angular/material' as mat;
@use 'palette';

@mixin theme($theme-or-color-config) {
  $color-config: mat.get-color-config($theme-or-color-config);
  $accent: map-get($color-config, accent);
  $selected-background-color: mat.get-color-from-palette($accent);
  $selected-hover-background-color: mat.get-color-from-palette($accent, darker);

  sg-text-marker {
    background-color: palette.$dark-text;
    color: white;

    &:after {
      border-color: palette.$dark-text transparent transparent;
    }

    &:hover {
      background-color: black;

      &:after {
        border-top-color: black;
      }
    }

    &.selected {
      background-color: $selected-background-color;
      box-shadow: 0 0 0 1px white;

      &:before {
        border-color: white transparent transparent;
      }

      &:after {
        border-top-color: $selected-background-color;
      }

      &:hover {
        background-color: $selected-hover-background-color;

        &:after {
          border-top-color: $selected-hover-background-color;
        }
      }
    }
  }
}
