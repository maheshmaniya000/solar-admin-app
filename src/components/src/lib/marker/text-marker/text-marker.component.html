{{ text }}
<svg
  *ngIf="!caretHidden"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  width="14"
  height="7"
  viewBox="0 0 14 7"
>
  <defs>
    <style>
      .b {
        fill: url(#a);
      }
    </style>
    <radialGradient
      id="a"
      cx="0.5"
      cy="0.5"
      r="0.5"
      gradientUnits="objectBoundingBox"
    >
      <stop offset="0" stop-opacity="0.396" />
      <stop offset="1" stop-opacity="0" />
    </radialGradient>
  </defs>
  <ellipse class="b" cx="7" cy="3.5" rx="7" ry="3.5" />
</svg>
