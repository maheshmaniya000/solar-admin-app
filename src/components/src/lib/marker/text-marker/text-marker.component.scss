:host {
  position: relative;
  display: inline-flex;
  box-sizing: border-box;
  align-items: center;
  height: 14px;
  max-width: fit-content;
  padding: 2px;
  font-size: 11px;
  border-radius: 2px;
  white-space: nowrap;
  cursor: pointer;

  &.large {
    height: 17px;
    font-size: 13px;
  }

  &:not(.caretHidden) {
    &:after,
    &.selected::before {
      content: '';
      position: absolute;
      display: block;
      left: 50%;
      border-style: solid;
    }

    $after-caret-height: 4px;
    $before-caret-height: $after-caret-height + 1px;
    $after-caret-middle-width: 2.75px;
    $before-caret-middle-width: $after-caret-middle-width + 1px;

    @mixin caret($middle-width, $height) {
      bottom: -$height;
      margin-left: -$middle-width;
      border-width: $height $middle-width 0;
    }

    &:after {
      @include caret($after-caret-middle-width, $after-caret-height);
      z-index: 2;
    }

    &.selected::before {
      @include caret($before-caret-middle-width, $before-caret-height);
      z-index: 1;
    }
  }
}

svg {
  position: absolute;
  bottom: -8px;
  left: 50%;
  margin-left: -7px;
}
