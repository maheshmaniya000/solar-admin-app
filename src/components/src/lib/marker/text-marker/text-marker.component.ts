import {
  ChangeDetectionStrategy,
  Component,
  HostBinding,
  Input
} from '@angular/core';

import { Vector2 } from '@solargis/types/pvlib-api';

import { AbstractMarkerComponent } from '../abstract-marker.component';

@Component({
  selector: 'sg-text-marker',
  templateUrl: './text-marker.component.html',
  styleUrls: ['./text-marker.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [
    // eslint-disable-next-line no-use-before-define
    { provide: AbstractMarkerComponent, useExisting: TextMarkerComponent }
  ]
})
export class TextMarkerComponent extends AbstractMarkerComponent {
  @Input() text: string;

  @HostBinding('class.caretHidden')
  @Input()
  caretHidden = false;

  getPivotPoint(): Vector2 {
    const pivotPoint = super.getPivotPoint();
    return {
      u: pivotPoint.u,
      v: this.caretHidden ? pivotPoint.v / 2 : pivotPoint.v * 1.25
    };
  }
}
