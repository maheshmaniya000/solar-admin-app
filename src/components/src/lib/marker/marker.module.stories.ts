import { MatButtonModule } from '@angular/material/button';
import { moduleMetadata, Story } from '@storybook/angular';

import { SgDefaultsModule } from '../defaults/defaults.module';
import { MarkerModule } from './marker.module';

export default {
  title: 'Solargis/Components/Marker',
  decorators: [
    moduleMetadata({
      imports: [MarkerModule, MatButtonModule, SgDefaultsModule]
    })
  ]
};

export const singleSelect: Story = props => ({
  props,
  template: `
    <div class="container" data-test="single-select-marker-component-story">
      <div class="markers-grid">
        <span class="label">Empty icon marker</span>
        <sg-icon-marker [selected]="selected"></sg-icon-marker>
        <span class="label">Icon marker</span>
        <sg-icon-marker icon="latlon_16px" [selected]="selected"></sg-icon-marker>
        <span class="label">Text marker</span>
        <sg-text-marker text="abcd" [selected]="selected"></sg-text-marker>
        <span class="label">Large text marker</span>
        <sg-text-marker class="large" text="ab" [selected]="selected"></sg-text-marker>
        <span class="label">Text marker without caret</span>
        <sg-text-marker text="abcd" [selected]="selected" [caretHidden]="true"></sg-text-marker>
        <span class="label">Large text marker without caret</span>
        <sg-text-marker class="large" text="ab" [selected]="selected" [caretHidden]="true"></sg-text-marker>
      </div>
    </div>
    `,
  styles: [
    `
    .container {
      height: 100vh;
      background-color: #a5adba;
    }

    .markers-grid {
      display: grid;
      width: 270px;
      grid-template-columns: repeat(2, auto);
      grid-template-rows: repeat(4, auto);
      grid-gap: 35px;
      padding: 15px;
    }

    .label {
      align-self: center;
    }
    `
  ]
});

singleSelect.argTypes = {
  selected: {
    name: 'Selected',
    control: { type: 'radio', options: [true, false] },
    defaultValue: false
  }
};
