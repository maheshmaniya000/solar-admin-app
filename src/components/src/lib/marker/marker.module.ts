import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatIconModule } from '@angular/material/icon';

import { IconMarkerComponent } from './icon-marker/icon-marker.component';
import { TextMarkerComponent } from './text-marker/text-marker.component';

@NgModule({
  declarations: [IconMarkerComponent, TextMarkerComponent],
  exports: [IconMarkerComponent, TextMarkerComponent],
  imports: [MatIconModule, CommonModule]
})
export class MarkerModule {}
