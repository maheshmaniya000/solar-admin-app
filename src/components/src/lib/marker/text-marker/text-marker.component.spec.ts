import { ElementRef } from '@angular/core';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';

import { mockElementDimensions } from '../../utils/test';
import { MarkerModule } from '../marker.module';
import { TextMarkerComponent } from './text-marker.component';

import spyOn = jest.spyOn;

describe('TextMarkerComponent', () => {
  describe('unit', () => {
    let component: TextMarkerComponent;

    beforeEach(() => {
      const element = document.createElement('div');
      mockElementDimensions(element, 200, 40);
      component = new TextMarkerComponent(new ElementRef(element));
    });

    describe('getPivotPoint', () => {
      it(`should return the text marker's pivot point when showing caret`, () => {
        expect(component.getPivotPoint()).toEqual({ u: 100, v: 50 });
      });

      it(`should return center of element when not showing caret`, () => {
        component.caretHidden = true;
        expect(component.getPivotPoint()).toEqual({ u: 100, v: 20 });
      });
    });
  });

  describe('component', () => {
    let spectator: Spectator<TextMarkerComponent>;
    const createComponent = createComponentFactory({
      component: TextMarkerComponent,
      imports: [MarkerModule],
      declareComponent: false
    });

    beforeEach(() => (spectator = createComponent()));

    it('should render component with text & caret', () => {
      spectator.setInput('text', 'some text');
      expect(spectator.fixture).toMatchSnapshot();
    });

    it('should render component with text & without caret', () => {
      spectator.setInput('text', 'some text');
      spectator.setInput('caretHidden', true);
      expect(spectator.fixture).toMatchSnapshot();
    });

    it('should emit selectedChange event when clicked', () => {
      const emitSpy = spyOn(spectator.component.selectedChange, 'emit');
      spectator.click(spectator.element);
      expect(emitSpy).toHaveBeenCalledTimes(1);
      spectator.setInput('selected', true);
      spectator.click(spectator.element);
      expect(emitSpy).toHaveBeenCalledTimes(2);
    });

    it('should toggle selected class when selected property changes', () => {
      expect(spectator.element).not.toHaveClass('selected');
      spectator.setInput('selected', true);
      spectator.detectChanges();
      expect(spectator.element).toHaveClass('selected');
      spectator.setInput('selected', false);
      spectator.detectChanges();
      expect(spectator.element).not.toHaveClass('selected');
    });
  });
});
