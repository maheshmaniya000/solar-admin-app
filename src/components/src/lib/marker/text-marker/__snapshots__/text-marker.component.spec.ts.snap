// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextMarkerComponent component should render component with text & caret 1`] = `
<sg-text-marker
  __ngContext__={[Function LRootView]}
  caretHidden="false"
  elementRef={[Function ElementRef]}
  selected="false"
  selectedChange={[Function EventEmitter_]}
  text={[Function String]}
>
  some text
  <svg
    height="7"
    viewBox="0 0 14 7"
    width="14"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <defs>
      <style>
         .b { fill: url(#a); } 
      </style>
      <radialgradient
        cx="0.5"
        cy="0.5"
        gradientUnits="objectBoundingBox"
        id="a"
        r="0.5"
      >
        <stop
          offset="0"
          stop-opacity="0.396"
        />
        <stop
          offset="1"
          stop-opacity="0"
        />
      </radialgradient>
    </defs>
    <ellipse
      class="b"
      cx="7"
      cy="3.5"
      rx="7"
      ry="3.5"
    />
  </svg>
</sg-text-marker>
`;

exports[`TextMarkerComponent component should render component with text & without caret 1`] = `
<sg-text-marker
  __ngContext__={[Function LRootView]}
  caretHidden={[Function Boolean]}
  elementRef={[Function ElementRef]}
  selected="false"
  selectedChange={[Function EventEmitter_]}
  text={[Function String]}
>
  some text
</sg-text-marker>
`;
